src/main/java/graphql/GraphQL.java

```diff
@@ -280,50 +279,106 @@ public ExecutionResult execute(String query, String operationName, Object contex
 
-        InstrumentationContext<Document> parseCtx = instrumentation.beginParse(new InstrumentationExecutionParameters(query, operationName, context, variables));
         Parser parser = new Parser();
         Document document;
         try {
-            document = parser.parseDocument(query);
-            parseCtx.onEnd(document);
+            document = parser.parseDocument(executionInput.getQuery());
         } catch (ParseCancellationException e) {
-            RecognitionException recognitionException = (RecognitionException) e.getCause();
-            SourceLocation sourceLocation = null;
-            if (recognitionException != null) {
-                sourceLocation = new SourceLocation(recognitionException.getOffendingToken().getLine(), recognitionException.getOffendingToken().getCharPositionInLine());
-            }
-            InvalidSyntaxError invalidSyntaxError = new InvalidSyntaxError(sourceLocation);
-            return new ExecutionResultImpl(Collections.singletonList(invalidSyntaxError));
+            parseInstrumentation.onEnd(e);
+            return ParseResult.ofError((RecognitionException) e.getCause());
         }
 
-        InstrumentationContext<List<ValidationError>> validationCtx = instrumentation.beginValidation(new InstrumentationValidationParameters(query, operationName, context, variables, document));
+        parseInstrumentation.onEnd(document);
+        return ParseResult.of(document);
+    }
+
+    private List<ValidationError> validate(ExecutionInput executionInput, Document document) {
+        InstrumentationContext<List<ValidationError>> validationCtx = instrumentation.beginValidation(new InstrumentationValidationParameters(executionInput, document));
 
         Validator validator = new Validator();
         List<ValidationError> validationErrors = validator.validateDocument(graphQLSchema, document);
 
         validationCtx.onEnd(validationErrors);
+        return validationErrors;
+    }
 
-        if (validationErrors.size() > 0) {
-            return new ExecutionResultImpl(validationErrors);
-        }
-        ExecutionId executionId = idProvider.provide(query, operationName, context);
+    private ExecutionResult execute(ExecutionInput executionInput, Document document) {
+        String query = executionInput.getQuery();
+        String operationName = executionInput.getOperationName();
+        Object context = executionInput.getContext();
 
         Execution execution = new Execution(queryStrategy, mutationStrategy, subscriptionStrategy, instrumentation);
-        ExecutionResult result = execution.execute(executionId, graphQLSchema, context, root, document, operationName, variables);
+        ExecutionId executionId = idProvider.provide(query, operationName, context);
+        return execution.execute(document, graphQLSchema, executionId, executionInput);
```

Comment:
This `@Internal` API change (and corresponding changes in `Execution` and `ExecutionTest`) is new in this amended commit and was not previously reviewed.